apps/OboeTester/app/src/main/java/com/mobileer/oboetester/RoundTripLatencyActivity.java

```diff
@@ -165,30 +201,287 @@ public void start() {
+        }
+
+        private String reportResults(String prefix) {
+            String message;
+            message = prefix + "buffer.bursts.low = " + mLowBufferBursts + "\n";
+            message += prefix + "latency.frames.low = " + mLowBufferLatency + "\n";
+            message += prefix + "buffer.bursts.high = " + mHighBufferBursts + "\n";
+            message += prefix + "latency.frames.high = " + mHighBufferLatency + "\n";
+            message += prefix + "result = " + mMessage + "\n";
+            return message;
         }
     }
-    AverageLatencyTestRunner mAverageLatencyTestRunner = new AverageLatencyTestRunner();
+
+    protected class ScanLatencyTestRunner extends MultipleLatencyTestRunner {
+        BinaryDiscontinuityFinder inputFinder;
+        BinaryDiscontinuityFinder outputFinder;
+        private static final int MAX_BAD_RUNS_ALLOWED = 5; // arbitrary
+        private int mBadCount = 0; // number of bad measurements
+
+        private static final int STATE_SCANNING_OUTPUT = 0;
+        private static final int STATE_SCANNING_INPUT = 1;
+        private static final int STATE_DONE = 2;
+        private int mState = STATE_SCANNING_OUTPUT;
+
+        // Called on UI thread.
+        @Override
+        String onAnalyserDone() {
+            int result = BinaryDiscontinuityFinder.RESULT_OK;
+            String message = "";
+
+            if (!mActive) {
+                message = "done";
+            } else if (getMeasuredResult() != 0) {
+                mBadCount++;
+                if (mBadCount > MAX_BAD_RUNS_ALLOWED) {
+                    cancel();
+                    result = BinaryDiscontinuityFinder.RESULT_ERROR;
+                    updateButtons(false);
+                    message = "scanning cancelled due to error, " + mBadCount + " bad runs\n";
+                } else {
+                    message = "skipping this bad run, "
+                            + mBadCount + " of " + MAX_BAD_RUNS_ALLOWED + " max\n";
+                    result = mActualBufferBursts;
+                }
+            } else {
+                switch (mState) {
+                    case STATE_SCANNING_OUTPUT:
+                        outputFinder.setFramesPerBurst(mOutputFramesPerBurst);
+                        result = outputFinder.onAnalyserDone(getMeasuredLatency(),
+                                getMeasuredConfidence(),
+                                mActualBufferBursts,
+                                mOutputBufferCapacityInBursts,
+                                mOutputIsMMapExclusive);
+                        mBufferBursts = result;
+
+                        break;
+                    case STATE_SCANNING_INPUT:
+                        inputFinder.setFramesPerBurst(mInputFramesPerBurst);
+                        result = inputFinder.onAnalyserDone(getMeasuredLatency(),
+                                getMeasuredConfidence(),
+                                mInputMarginBursts,
```

Comment:
For my understanding, why does input use input margin instead of buffer bursts? What is the difference?